packages/blockchain/src/index.ts

```diff
@@ -498,71 +508,115 @@ export default class Blockchain implements BlockchainInterface {
+        // See if there is a new majority consensus to update the signer list
         const lastEpochBlockNumber = header.number.sub(
           header.number.mod(new BN(this._common.consensusConfig().epoch))
         )
-        const beneficiaryVotesAuth = this._cliqueLatestVotes.filter(
-          (vote) =>
+        const limit = this.cliqueSignerLimit()
+        let activeSigners = this.cliqueActiveSigners()
+        let consensus = false
+
+        // AUTH vote analysis
+        let votes = this._cliqueLatestVotes.filter((vote) => {
+          return (
             vote[0].gte(lastEpochBlockNumber) &&
+            !vote[1][0].equals(signer) &&
             vote[1][1].equals(beneficiary) &&
             vote[1][2].equals(CLIQUE_NONCE_AUTH)
-        )
-        const beneficiaryVotesDrop = this._cliqueLatestVotes.filter(
-          (vote) =>
+          )
+        })
+        const beneficiaryVotesAUTH: Address[] = []
+        for (const vote of votes) {
+          const num = beneficiaryVotesAUTH.filter((voteCMP) => {
+            return voteCMP.equals(vote[1][0])
+          }).length
+          if (num === 0) {
+            beneficiaryVotesAUTH.push(vote[1][0])
+          }
+        }
+        let numBeneficiaryVotesAUTH = beneficiaryVotesAUTH.length
+        if (round === 2 && nonce.equals(CLIQUE_NONCE_AUTH)) {
+          numBeneficiaryVotesAUTH += 1
+        }
+        // Majority consensus
+        if (numBeneficiaryVotesAUTH >= limit) {
+          consensus = true
+          // Authorize new signer
+          activeSigners.push(beneficiary)
+          activeSigners.sort((a, b) => {
+            // Sort by buffer size
+            return a.toBuffer().compare(b.toBuffer())
+          })
+          // Discard votes for added signer
+          this._cliqueLatestVotes = this._cliqueLatestVotes.filter(
+            (vote) => !vote[1][1].equals(beneficiary)
+          )
+          debug(
+            `[Block ${header.number.toNumber()}] Clique majority consensus (AUTH ${beneficiary})`
+          )
+        }
+        // DROP vote
+        votes = this._cliqueLatestVotes.filter((vote) => {
+          return (
             vote[0].gte(lastEpochBlockNumber) &&
+            !vote[1][0].equals(signer) &&
             vote[1][1].equals(beneficiary) &&
             vote[1][2].equals(CLIQUE_NONCE_DROP)
-        )
-        const limit = this.cliqueSignerLimit()
-        const consensus =
-          beneficiaryVotesAuth.length >= limit || beneficiaryVotesDrop.length >= limit
-        const auth = beneficiaryVotesAuth.length >= limit
+          )
+        })
+        const beneficiaryVotesDROP: Address[] = []
+        for (const vote of votes) {
+          const num = beneficiaryVotesDROP.filter((voteCMP) => {
+            return voteCMP.equals(vote[1][0])
+          }).length
+          if (num === 0) {
+            beneficiaryVotesDROP.push(vote[1][0])
+          }
+        }
+        let numBeneficiaryVotesDROP = beneficiaryVotesDROP.length
```

Comment:
@ryanio thanks for review, wasn't so sure about here how to do this elegantly, with a direct additional `filter()` or `reduce()` call, if you know a better way feel free to rewrite on some subsequent PR (then we really make sure that it still syncs these first 6.000 or so Goerli blocks though, a bit trick. I was wondering if we should extract the case which was failing for us before the rewrite and add (propose) as an additional test case to the EIP. I am planning to do a PR with some suggestions for improvements and clarifications anyhow once we are stable (might be more or less: now, we'll see, likely though), have taken notes along the lines. If you have something to add feel free to send me. We can alternatively also fork the EIP repo to `ethereumjs` and do the PR together).